File: smaKeeper.js
const { SlidingWindowArr } = require('./slidingWindowArr');
const { sum } = require('./common');

class SmaKeeper {
  currentSma = 0;
  constructor(period) {
    this.period = period;
    this.historyValues = new SlidingWindowArr({ maxLen: this.period });
  }

  add(val) {
    if (this.historyValues.length() === 0) {
      this.historyValues.push(val);
      this.currentSma = val;
    } else {
      if (this.historyValues.length() >= this.period) {
        // we need to reduce the first value and add the last value
        this.currentSma -= this.historyValues.get(0) / this.period;
        this.currentSma += val / this.period;
        this.historyValues.push(val);
      } else {
        // the length hasn't reached max, we can just add to new sma
        this.historyValues.push(val);
        this.currentSma = sum(this.historyValues.toUnorderedArr()) / this.historyValues.length();
      }
    }

    return this.currentSma;
  }

  peekNext(val) {
    if (this.historyValues.length() === 0) {
      return val;
    } else {
      if (this.historyValues.length() >= this.period) {
        let tmpSma = this.currentSma;
        tmpSma -= this.historyValues.get(0) / this.period;
        tmpSma += val / this.period;
        return tmpSma;
      } else {
        // the length hasn't reached max
        return val;
      }
    }
  }

  getValues() {
    return this.historyValues;
  }

  get() {
    if (this.historyValues.length() === 0) {
      throw new Error(`SmaKeeper cannot get sma when there is no value added yet`);
    }
    return this.currentSma;
  }

  length() {
    return this.historyValues.length();
  }

  getPeriod() {
    return this.period;
  }
}

module.exports = { SmaKeeper }

File: nan/smaKeeper.cc
#include "smaKeeper.h"
#include "slidingWindowArr.h"
#include "slidingWindowArr.cc"
#include <numeric>

SmaKeeper::SmaKeeper(int period): _slidingWindowArr(period)
 {

  this->_sma = 0.0;
  this->_period = period;
}

void SmaKeeper::add(float price)
{

  if (this->_slidingWindowArr.length() < this->_period)
  {
    this->_slidingWindowArr.push(price);
    float sum = 0;
    for (int i = 0; i < this->_slidingWindowArr.length(); i++)
    {
      sum += this->_slidingWindowArr.get(i);
    }
    this->_sma = sum / this->_slidingWindowArr.length();
  }
  else
  {
    float removedValue = this->_slidingWindowArr.first();

    this->_slidingWindowArr.push(price);

    this->_sma = this->_sma - removedValue / this->_period + price / this->_period;
  }
}

float SmaKeeper::get() const
{
  return this->_sma;
}
